mypy/specialtype.py

```diff
@@ -0,0 +1,912 @@
+from collections import OrderedDict
+
+from typing import List, Dict, Tuple, cast, Optional, Callable, TYPE_CHECKING
+
+from mypy.exprtotype import expr_to_unanalyzed_type, TypeTranslationError
+from mypy.nodes import (
+    TypeInfo, AssignmentStmt, FuncDef, ClassDef, Var, GDEF, Expression,
+    Block, NameExpr, TupleExpr, ListExpr, ExpressionStmt, PassStmt,
+    DictExpr, CallExpr, RefExpr, Context, SymbolTable, UNBOUND_TVAR,
+    MDEF, Decorator, TypeVarExpr, NewTypeExpr, StrExpr, BytesExpr,
+    ARG_POS, ARG_NAMED, ARG_NAMED_OPT, NamedTupleExpr, TypedDictExpr, Argument,
+    UnicodeExpr, EllipsisExpr, TempNode, EnumCallExpr,
+    COVARIANT, CONTRAVARIANT, INVARIANT, ARG_OPT, SymbolTableNode
+)
+from mypy.types import (
+    NoneTyp, CallableType, Instance, Type, TypeVarType, AnyType,
+    TypeVarDef, TupleType, UnboundType, TypedDictType, TypeType,
+)
+from mypy import join
+if TYPE_CHECKING:
+    import mypy.semanal
+
+
+class Special:
+    """
+    Groups special-cased types:
+    * NamedTuple
```

Comment:
I wonder if it would make sense to move the code related to various special cased types to separate classes (or top-level functions), to improve cohesion? They could all live in the same file still. It may make sense to have a single entry point to the module though. (Note that it's probably not worth using inheritance.)